src/Http/Http/src/Internal/ResponseCookies.cs

```diff
@@ -88,6 +88,57 @@ public void Append(string key, string value, CookieOptions options)
             Headers[HeaderNames.SetCookie] = StringValues.Concat(Headers[HeaderNames.SetCookie], cookieValue);
         }
 
+        /// <inheritdoc />
+        public void Append(IDictionary<string, string> keyValuePairs, CookieOptions options)
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            // SameSite=None cookies must be marked as Secure.
+            if (!options.Secure && options.SameSite == SameSiteMode.None)
+            {
+                if (_logger == null)
+                {
+                    var services = _features.Get<Features.IServiceProvidersFeature>()?.RequestServices;
+                    _logger = services?.GetService<ILogger<ResponseCookies>>();
+                }
+
+                if (_logger != null)
+                {
+                    foreach (var keyValuePair in keyValuePairs)
+                    {
+                        Log.SameSiteCookieNotSecure(_logger, keyValuePair.Key);
+                    }
+                }
+            }
+
+            var setCookieHeaderValue = new SetCookieHeaderValue(string.Empty)
+            {
+                Domain = options.Domain,
+                Path = options.Path,
+                Expires = options.Expires,
+                MaxAge = options.MaxAge,
+                Secure = options.Secure,
+                SameSite = (Net.Http.Headers.SameSiteMode)options.SameSite,
+                HttpOnly = options.HttpOnly
+            };
+
+            var cookierHeaderValue = setCookieHeaderValue.ToString()[1..];
+            var cookies = new string[keyValuePairs.Count];
```

Comment:
I think you could borrow this from the ArrayPool. Alternatively you could simply call `Append` n times. It's effectively what the `IHeaderDictionary.Append(IList<T>)` seems to do: https://github.com/dotnet/aspnetcore/blob/5b3ba86121c6afc27ac3851308e0f00b8c809600/src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs#L139-L144

Reply:
In respect to `ArrayPool`:

It sounded cool so I tried it using ` ArrayPool<string>.Shared.Rent(keyValuePairs.Count())` but it actually increases the `mean` about 100ns-200ns.

----

In respect to `Headers.Append`:
 
https://github.com/dotnet/aspnetcore/blob/9a5538162a903dc16d9de37585696a45ca4ce32d/src/Http/Http/src/Internal/ResponseCookies.cs#L133-L139

I am indeed calling it but just once at the end, I just wanted to avoid calling  `Headers.Append` multiple times because at the end it calls `StringValues.concat` and it means unnecessary allocation of `arrays`, hopefully I was able to build the full array of new cookies just like the example you gave me.
  
https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/Http/Http.Abstractions/src/Internal/ParsingHelpers.cs#L160-L161

Thanks!